gcp/warehouse.go

```diff
@@ -0,0 +1,160 @@
+
+// Package gcp abstracts interacting with certain aspects of Google Cloud
+// Platform, such as creating service account keys and access tokens.
+package gcp
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/hex"
+	"fmt"
+	"net/http"
+	"path"
+
+	"golang.org/x/crypto/sha3"
+	cloudresourcemanager "google.golang.org/api/cloudresourcemanager/v1"
+	"google.golang.org/api/googleapi"
+	iam "google.golang.org/api/iam/v1"
+	iamcredentials "google.golang.org/api/iamcredentials/v1"
+)
+
+// AccountWarehouseOptions is used with NewWarehouse to configure a new warehouse.
+type AccountWarehouseOptions struct {
+	Project     string
+	DefaultRole string
+	Scopes      []string
+}
+
+// AccountWarehouse is used to create Google Cloud Platform Service Account
+// keys and access tokens associated with a specific identity.
+type AccountWarehouse struct {
+	opts  AccountWarehouseOptions
+	iam   *iam.Service
+	creds *iamcredentials.Service
+	crm   *cloudresourcemanager.Service
+}
+
+// NewAccountWarehouse creates a new AccountWarehouse using the provided client
+// and options.
+func NewAccountWarehouse(client *http.Client, opts *AccountWarehouseOptions) (*AccountWarehouse, error) {
+	iamSvc, err := iam.New(client)
+	if err != nil {
+		return nil, fmt.Errorf("creating IAM client: %v", err)
+	}
+
+	creds, err := iamcredentials.New(client)
+	if err != nil {
+		return nil, fmt.Errorf("creating IAM credentials client: %v", err)
+	}
+
+	crm, err := cloudresourcemanager.New(client)
+	if err != nil {
+		return nil, fmt.Errorf("creating cloud resource manager client: %v", err)
+	}
+
+	return &AccountWarehouse{
+		opts:  *opts,
+		iam:   iamSvc,
+		creds: creds,
+		crm:   crm,
+	}, nil
+}
+
+// GetAccountKey returns a service account key associated with id.
+func (wh *AccountWarehouse) GetAccountKey(ctx context.Context, id string) ([]byte, error) {
+	account, err := wh.getBackingAccount(ctx, id)
+	if err != nil {
+		return nil, fmt.Errorf("getting backing account: %v", err)
+	}
+
+	// TODO(#2): Remove old keys from service account before creating new keys.
+	keys := wh.iam.Projects.ServiceAccounts.Keys
+	result, err := keys.Create(accountID("-", account), &iam.CreateServiceAccountKeyRequest{
+		PrivateKeyType: "TYPE_GOOGLE_CREDENTIALS_FILE",
+	}).Context(ctx).Do()
+	if err != nil {
+		return nil, fmt.Errorf("creating key: %v", err)
+	}
+
+	out, err := base64.StdEncoding.DecodeString(result.PrivateKeyData)
+	if err != nil {
+		return nil, fmt.Errorf("decoding key: %v", err)
+	}
+
+	return out, nil
+}
+
+// GetAccessToken returns an access token for the service account uniquely
+// associated with id.
+func (wh *AccountWarehouse) GetAccessToken(ctx context.Context, id string) (string, error) {
+	account, err := wh.getBackingAccount(ctx, id)
+	if err != nil {
+		return "", fmt.Errorf("getting backing account: %v", err)
+	}
+
+	response, err := wh.creds.Projects.ServiceAccounts.GenerateAccessToken(accountID("-", account), &iamcredentials.GenerateAccessTokenRequest{
```

Comment:
Why "-" in the call to accountID in a couple places?

Reply:
@kemp-google Because that's the definition as noted at the following line:

https://github.com/googleapis/google-api-go-client/blob/master/iamcredentials/v1/iamcredentials-gen.go#L114

Reply:
Or in the doc itself:  

https://godoc.org/google.golang.org/api/iamcredentials/v1#GenerateIdTokenRequest